TCP socket events for EBPF

haesbaert opened the pull request:
This implements basic socket events for TCP, it tracks
connect(2)/accept(2)/close(2) on respective sockets as well as builds the
initial snapshot of existing sockets by parsing /proc.

Sockets are expressed in quark_socket{} and correlate with processes. Since a
socket might be in more than one process, two fields are provided:
 - pid_origin, the pid that established the connection.
 - pid_last_use, the last pid that affected it, if another process did the final
   close, pid_origin remains the same, but pid_last_use points to the last one.

We stash the lookup of quark_process_lookup(pid_origin) in the event process
pointer, which *can* be NULL, and the socket itself is stashed in the socket
member.

quark_event{} might need some adjustments in the future.

The EBPF probes are a bit subpar and they currently have two issues:
 - A connection that never sent a byte doesn't produce an exit event.
   elastic/ebpf#217
 - Exit events happen before TCP close is issued
   elastic/ebpf#216

I don't think this is good enough and will consider it experimental until both
issues are solved, first one has a fix on the way, second one likely involves
rewriting the whole probe.

Quark sockets don't produce snapshot events, because snapshot will be removed
and the user will have to use the iterator if desired.

In order to test this, pass -S to quark-mon(8).

Co-authored-by: Nicholas Berlin <56366649+nicholasberlin@users.noreply.github.com>

nicholasberlin commented on quark.c:
```diff
@@ -408,43 +479,179 @@ process_cache_inherit(struct quark_queue *qq, struct quark_process *qp, int ppid
+
+static struct quark_socket *
+socket_cache_get(struct quark_queue *qq,
+    struct quark_sockaddr *local, struct quark_sockaddr *remote, int alloc)
```
can local and remote be const?

haesbaert replied:
Could, as it's not an external facing function, I wouldn't care much (most of the other code is node const)

nicholasberlin commented on quark.c:
```diff
@@ -408,43 +479,179 @@ process_cache_inherit(struct quark_queue *qq, struct quark_process *qp, int ppid
+}
+
+const struct quark_socket *
+quark_socket_lookup(struct quark_queue *qq,
```
I don't see this function being called. Enlighten me please.

nicholasberlin replied:
Oh, I suppose these are used by library users.

haesbaert replied:
Yep!

nicholasberlin commented on quark.c:
```diff
@@ -1329,7 +1588,102 @@ sproc_namespace(struct quark_process *qp, const char *path, u32 *dst, int dfd)
+		    d->d_name, buf, inode, ss);
+	}
+
+	/* closedir() closes the backing `fdfd` */
```
I was wondering that. Good comment.

nicholasberlin commented on quark.c:
```diff
@@ -1329,7 +1588,102 @@ sproc_namespace(struct quark_process *qp, const char *path, u32 *dst, int dfd)
+		    &qsk_key);
+		if (qsk != NULL) {
+			/* XXX TODO WRITE AN EXAMPLE TO TRIGGER THIS */
+			printf("EXISTING !!! pid %d fd %s -> %s (inode=%lu, ss=%p)\n", pid,
```
Is this an example of a process forking without closing the socket?

haesbaert replied:
Yep, or sending the socket via sendmsg() to another process, but basically any case where a socket is referenced/used by more than one process.

nicholasberlin commented on quark.c:
```diff
@@ -1362,12 +1716,214 @@ sproc_pid(struct quark_queue *qq, int pid, int dfd)
+	if (state != TCP_ESTABLISHED && state != TCP_CLOSE_WAIT)
+		return (0);
+
+#if 1
```
still need this debugging?

haesbaert replied:
gonna zap it

nicholasberlin commented on quark.c:
```diff
@@ -1756,6 +2344,7 @@ quark_queue_default_attr(struct quark_queue_attr *qa)
 	bzero(qa, sizeof(*qa));
 
 	qa->flags = QQ_ALL_BACKENDS;
+	qa->flags |= QQ_NET;	/* XXX REMOVE ME XXX */
```
reminder

nicholasberlin commented on quark.c:
```diff
@@ -1996,6 +2590,57 @@ quark_queue_get_snap_event(struct quark_queue *qq)
+		break;
+	default:
+		qwarnx("invalid conn->conn %d\n", conn->conn);
+		return (NULL);
```
if `qsk` was allocated by `socket_cache_get`, should it be freed here?

haesbaert replied:
Yep, this is more like a programming error, but we should be correct anyway, I'll have to change this function a bit, because there is a second issue:
Basically socket_cache_get() cannot do the "lookup or alloc", it needs to be split into two. When we get an ACCEPT/CONNECT, we must hit the alloc case, which might incur us evicting the old entry and creating a new one.
This can be the case if, for any reason, we missed the close event.

haesbaert replied:
In the upcoming version, it doesn't have qsk allocated on that path anymore, so this is gone.

haesbaert commented on quark.c:
```diff
@@ -408,43 +479,179 @@ process_cache_inherit(struct quark_queue *qq, struct quark_process *qp, int ppid
 
 static int
-process_cache_gc(struct quark_queue *qq)
+process_by_pid_cmp(struct quark_process *a, struct quark_process *b)
```
They could, but all the other _cmp are not, so I'd just keep the same.
Generally I'm more worried about constifying external calls

haesbaert commented on quark.c:
```diff
@@ -408,43 +479,179 @@ process_cache_inherit(struct quark_queue *qq, struct quark_process *qp, int ppid
+ * Socket stuff
+ */
+
+#if 0
```
zapped them

haesbaert commented:
Pushed the proposed final version.
Relevant changes:
 - Renamed connection to sock
 - Reworked socket allocation and insertion
 - Addressed the review comments
 - Updated the commit message and this PR.
